code/modules/projectiles/projectile.dm

```diff
@@ -120,20 +134,76 @@
 /obj/item/projectile/Process_Spacemove(var/movement_dir = 0)
 	return 1 //Bullets don't drift in space
 
-
-/obj/item/projectile/proc/fire()
-	set waitfor = 0
-	while(loc)
-		if(!paused)
-			if((!( current ) || loc == current))
-				current = locate(Clamp(x+xo,1,world.maxx),Clamp(y+yo,1,world.maxy),z)
-			step_towards(src, current)
-			if(original && (original.layer>=2.75) || ismob(original))
-				if(loc == get_turf(original))
-					if(!(original in permutated))
-						Bump(original, 1)
-			Range()
-		sleep(1)
+/obj/item/projectile/proc/fire(var/setAngle)
+	if(setAngle)
+		Angle = setAngle
+	if(!legacy) //new projectiles
+		set waitfor = 0
+		while(loc)
+			if(!paused)
+				if((!( current ) || loc == current))
+					current = locate(Clamp(x+xo,1,world.maxx),Clamp(y+yo,1,world.maxy),z)

```

Comment:
This seems like it can be a intensive if. Does this fire every time a new projectile is made?